pkg/operator/csi/controller/driver/sync.go

```diff
@@ -0,0 +1,321 @@
+				status.ReadyReplicas += deployment.Status.UpdatedReplicas
+			}
+		}
+	}
+
+	// Finally, set the conditions
+
+	// The operator does not have any prerequisites (at least now)
+	v1helpers.SetOperatorCondition(&status.Conditions,
+		opv1.OperatorCondition{
+			Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypePrereqsSatisfied),
+			Status: opv1.ConditionTrue,
+		})
+
+	// The operator is always upgradeable (at least now)
+	v1helpers.SetOperatorCondition(&status.Conditions,
+		opv1.OperatorCondition{
+			Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeUpgradeable),
+			Status: opv1.ConditionTrue,
+		})
+
+	// The operator is avaiable for now
+	v1helpers.SetOperatorCondition(&status.Conditions,
+		opv1.OperatorCondition{
+			Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeAvailable),
+			Status: opv1.ConditionTrue,
+		})
+
+	// Make it not available if daemonSet hasn't deployed the pods
+	if !isDaemonSetAvailable(daemonSet) {
+		v1helpers.SetOperatorCondition(&status.Conditions,
+			opv1.OperatorCondition{
+				Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeAvailable),
+				Status:  opv1.ConditionFalse,
+				Message: "Waiting for the DaemonSet to deploy the CSI Node Service",
+				Reason:  "AsExpected",
+			})
+	}
+
+	// Make it not available if deployment hasn't deployed the pods
+	if c.controllerManifest != nil {
+		if !isDeploymentAvailable(deployment) {
+			v1helpers.SetOperatorCondition(&status.Conditions,
+				opv1.OperatorCondition{
+					Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeAvailable),
+					Status:  opv1.ConditionFalse,
+					Message: "Waiting for Deployment to deploy the CSI Controller Service",
+					Reason:  "AsExpected",
+				})
+		}
+	}
+
+	// The operator is not progressing for now
+	v1helpers.SetOperatorCondition(&status.Conditions,
+		opv1.OperatorCondition{
+			Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeProgressing),
+			Status: opv1.ConditionFalse,
+			Reason: "AsExpected",
+		})
+
+	isProgressing, msg := c.getDaemonSetProgress(status, daemonSet)
+	if isProgressing {
+		v1helpers.SetOperatorCondition(&status.Conditions,
+			opv1.OperatorCondition{
+				Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeProgressing),
+				Status:  opv1.ConditionTrue,
+				Message: msg,
+				Reason:  "AsExpected",
+			})
+	}
+
+	if c.controllerManifest != nil {
+		// CSI Controller deployed, let's check its progressing state
+		isProgressing, msg := c.getDeploymentProgress(status, deployment)
+		if isProgressing {
+			v1helpers.SetOperatorCondition(&status.Conditions,
+				opv1.OperatorCondition{
+					Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeProgressing),
+					Status:  opv1.ConditionTrue,
+					Message: msg,
+					Reason:  "AsExpected",
+				})
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) getExpectedDeployment(spec *opv1.OperatorSpec) *appsv1.Deployment {
```

Comment:
One of the things we regretted about our early controllers was requiring particular shapes to the deployments.  We found that if we instead manipulated strings of yaml, we could do simple `sed s/orig/new/g` like replacements that ended up more flexible.

Notable wins about the `sed` approach
1. no special names on containers
2. no pre-determined substitution locations
3. original yaml does not need to parse into a deployment or daemonset (this is big for integer replacements for instance)
4. it's easy to wrap commands inside of bash or change the order of things.

And in your case it would allow a mutation function to come in as part of the `WithDeployment` function that accepts the bytes and spec if someone needs additional customization.  You would probably run your `sed` after the customization.

Reply:
Will address this next

Reply:
Done. The mutation function part is a work-in-progress, I'll add that after this is merged.